File: substring_test.cc
#include <gtest/gtest.h>
#include "substring.h"

TEST(SubstringTest, BasicAssertions) {
    EXPECT_EQ(wob::find("a", "a"), std::make_pair(0ul, true));
    EXPECT_EQ(wob::find("a", "ba"), std::make_pair(1ul, true));
    EXPECT_EQ(wob::find("abcd", "abcabcdabcde"), std::make_pair(3ul, true));
    EXPECT_EQ(std::get<1>(wob::find("a", "b")), false);
}

TEST(SubstringTest, EmptyStrings) {
    EXPECT_EQ(wob::find("", "a"), std::make_pair(0ul, true));
    EXPECT_EQ(std::get<1>(wob::find("a", "")), false);
}

File: test/substring_test.clj
(ns substring-test
  (:require [substring :as ss]
            [clojure.spec.alpha :as s]
            [clojure.test.check.generators :as gen]
            [clojure.spec.test.alpha :as stest]))

(use 'clojure.test)

(defn instr-fns [f]
  (stest/instrument `ss/strpos)
  (f)
  (stest/unstrument `ss/strpos))

(use-fixtures :once instr-fns)

(def str-gen #(gen/tuple
               (gen/fmap clojure.string/join
                         (gen/vector
                          (gen/elements ["a" "b"])))
               (gen/fmap clojure.string/join
                         (gen/vector
                          (gen/elements ["a" "b" "c"])))))

(defn pass? [t]
  (every? #(not (:failure %)) t))

(deftest check-strpos
  (is (and
       (pass? (stest/check `ss/strpos {:gen {:substring/find-args str-gen}}))
       (pass? (stest/check `ss/strpos)))))

(deftest basic
  (is (= (ss/strpos "abcd" "abcabcdabcde") 3))
  (is (= (ss/strpos "abc" "abcabcdabcde") 0))
  (is (= (ss/strpos "z" "abcabcdabcde") -1)))

(deftest empty-strings
  (is (= (ss/strpos "" "") 0))
  (is (= (ss/strpos "a" "") -1))
  (is (= (ss/strpos "" "a") 0)))

(deftest null-inputs
  (is (thrown? clojure.lang.ExceptionInfo (ss/strpos nil nil) -1)))
